torchaudio/datasets/datasets.py

```diff
@@ -0,0 +1,419 @@
+import os
+import pickle
+import random
+from functools import partial, reduce
+from warnings import warn
+
+import torch
+
+import torchaudio
+
+
+class Cache:
+    """
+    Wrap a generator so that, whenever a new item is returned, it is saved to disk in a pickle.
+    """
+
+    def __init__(self, generator, location):
+        self.generator = generator
+        self.location = location
+
+        self._id = id(self)
+        self._cache = []
+        self._internal_index = 0
+
+    def __iter__(self):
+        self._internal_index = 0
+        return self
+
+    def __next__(self):
+        if self._internal_index < len(self):
+            item = self[self._internal_index]
+        else:
+            item = next(self.generator)
+
+            file = str(self._id) + "-" + str(len(self))
+            file = os.path.join(self.location, file)
+            self._cache.append(file)
+
+            os.makedirs(self.location, exist_ok=True)
+            with open(file, "wb") as file:
+                pickle.dump(item, file)
+
+        self._internal_index += 1
+        return item
+
+    def __getitem__(self, index):
+        file = self._cache[index]
+        with open(file, "rb") as file:
+            item = pickle.load(file)
+        return item
+
+    def __len__(self):
+        # Return length of cache
+        return len(self._cache)
+
+
+class Buffer:
+    """
+    Wrap a generator so as to keep the last few in memory.
+    """
+
+    def __init__(self, generator, capacity=10):
+        self.generator = generator
+        self.capacity = capacity
+        self._cache = []
+        self._fill()
+
+    def _fill(self):
+        while len(self._cache) <= self.capacity:
+            self._cache.append(next(self.generator))
+
+    def __getitem__(self, n):
+        self._fill()
+        return self._cache[n]
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        item = self._cache.pop(0)
+        self._fill()
+        return item
+
+
+def download(urls, root_path):
```

Comment:
If you want to write an abstraction that explicitly downloads a file to disk I'd call this "download_to_file".

In general a generic download function that returns a stream is a bit tricky, because if a user doesn't consume the stream the connection might break down (?). Also when writing a file straight to disk from the web we can potentially bypass system memory.

However, having something that does indeed give you a stream of data from the web is very useful.

If you want to get fancy you could also try to auto-discover the filename to be used as in here: https://github.com/pytorch/text/blob/fd31bf3722e17dfd5998b8c4f32c0431f3016d59/torchtext/utils.py#L33